### PR TITLE
Added filters to search from filterable attributes. #6543

### PR DESCRIPTION
**Pull request by PritishC**
* Added a test for filtering and facet tally updation on filtering. Filters are applied to the facets as well.
* If an attribute is filterable, additional settings can be configured for it, such as displaying the facet counts, multi-selection of attribute values, etc.
* Added a boolean parameter to prevent faceting on autocomplete request.
* Added a template filter to append display_count values to facets.
* Updated copyright dates.

For reference: [SO Thread](http://stackoverflow.com/questions/12399326/convert-the-facet-filter-query-into-pyes-format)

### CODE REVIEW

**Comment by PritishC** [comment]
#13 needs to be merged and then will rebase this PR.

File: product.py [hunk 1]
```diff
@@ -153,7 +167,91 @@ def get_elastic_search_query(cls, search_phrase):
         )
 
     @classmethod
-    def search_on_elastic_search(
+    def _build_es_filter(cls, request, filterable_attributes=None):
```
Comment (tarunbhardwaj) [inline product.py]: Request is already imported, need not to pass

File: product.py [hunk 2]
```diff
@@ -184,25 +282,38 @@ def search_on_elastic_search(
...
-        # Build the query object
-        query = cls.get_elastic_search_query(search_phrase)
+        filterable_attributes = cls.get_filterable_attributes()
 
-        # Wrap it in a `~pyes.query.Search` object for convenience, by calling
-        # its search() method.
-        search_obj = query.search()
+        # Create the filter.
+        main_filter = cls._build_es_filter(
+            request, filterable_attributes=filterable_attributes
```
Comment (tarunbhardwaj) [inline product.py]: no request

File: product.py [hunk 3]
```diff
@@ -184,25 +282,38 @@ def search_on_elastic_search(
...
         """
-        # Build the query object
-        query = cls.get_elastic_search_query(search_phrase)
+        filterable_attributes = cls.get_filterable_attributes()
 
-        # Wrap it in a `~pyes.query.Search` object for convenience, by calling
-        # its search() method.
-        search_obj = query.search()
+        # Create the filter.
+        main_filter = cls._build_es_filter(
```
Comment (tarunbhardwaj) [inline product.py]: call this es_filter not main_filter

File: product.py [hunk 4]
```diff
@@ -184,25 +282,38 @@ def search_on_elastic_search(
...
+            request, filterable_attributes=filterable_attributes
+        )
 
-        # If the autocomplete handler wasn't the one to send a request,
-        # faceting can be done.
+        # Generate the `~pyes.query.Query` object.
+        query = cls._build_es_query(search_phrase)
+
+        # Now wrap the query in a `~pyes.query.Search` object for convenience.
+        # Apply any kwargs such as filters.
```
Comment (tarunbhardwaj) [inline product.py]: no kwargs now

File: product.py [hunk 5]
```diff
@@ -184,25 +282,38 @@ def search_on_elastic_search(
...
+        search_obj = query.search(filter=main_filter)
+
+        # Add faceting. Apply any kwargs such as filters.
```
Comment (tarunbhardwaj) [inline product.py]: no kwargs

File: product.py [hunk 6]
```diff
@@ -88,23 +93,32 @@ def get_filterable_attributes(cls):
         return Attribute.search([('filterable', '=', True)])
 
     @classmethod
-    def add_faceting_to_query(cls, search_obj):
+    def _build_es_facets(
```
Comment (tarunbhardwaj) [inline product.py]: as this method is updating in search_obj only better call this method `_update_es_facets`

File: product.py [hunk 7]
```diff
@@ -88,23 +93,32 @@ def get_filterable_attributes(cls):
...
-            search_obj.facet.add_term_facet(attribute)
+            search_obj.facet.add_term_facet(
+                attribute.name,
+                facet_filter=facet_filter,
+                all_terms=attribute.multiselect,
+                size=attribute.display_size,
+                order=attribute.display_order
+            )
 
         return search_obj
```
Comment (tarunbhardwaj) [inline product.py]: no need of returning search_obj is is just updating the same.

File: website.py [hunk 8]
```diff
@@ -60,4 +60,8 @@ def quick_search(cls):
             (phrase, products.count)
         )
 
-        return render_template('search-results.jinja', products=products)
+        return render_template(
+            'search-results.jinja',
+            products=products,
+            facets=products.result_set.facets
```
Comment (tarunbhardwaj) [inline website.py]: here you can pass one more arg for all the attributes which have display_count True. so in template you can just check if count need to display or not

**Comment by tarunbhardwaj** [comment]
LGTM

File: tests/test_pagination.py [hunk 9]
```diff
@@ -2,7 +2,7 @@
 """
     tests/test_pagination.py
 
-    :copyright: (c) 2014 by Openlabs Technologies & Consulting (P) Limited
+    :copyright: (c) 2015 by Openlabs Technologies & Consulting (P) Limited
```
Comment (tarunbhardwaj) [inline tests/test_pagination.py]: 2014-2015

File: product.py [hunk 10]
```diff
@@ -88,23 +93,30 @@ def get_filterable_attributes(cls):
...
         adds appropriate terms to the `facet` attribute of this object. By
         default, term facets are generated over filterable attributes.
         """
-        filterable_attributes = map(
-            lambda x: x.name, cls.get_filterable_attributes()
-        )
+        # If no filterable attributes in database, return search_obj without
+        # doing the faceting.
+        if not filterable_attributes:
+            return search_obj
```
Comment (tarunbhardwaj) [inline product.py]: No need to return

**Pull request merged**